azavea/climate-change-lab
# Feature/serialize ui state

## Overview

Update the Lab application to serialize project state to window.localStorage when the LabComponent is destroyed

### Notes

This is a pretty big refactor, had to touch alot of the internals.

Items of note:
- Rmoved ChartContainerComponent as it was an unnecessary layer of indirection that wasn't used for anything
- Refactor querying for ClimateModel/Scenario to separate services and simplified the public API
- Added the Chart and Project view models, and made their respective components use these instead of properties on the component (this made serialization much easier)
- Moved the chart loading spinner so that the chart settings div remains visible (only the chart body where the drawing happens is replaced by the spinner)

Future TODOs (all of these are separate tasks I think):
- Figure out how to get project to save if page is refreshed (state is only saved currently when the LabComponent is destroyed, e.g. on a state change back to dashboard). I don't know what the hook for this is.
- Make each chart load its own data. This should be easier now too with the removal of ChartContainerComponent and the refactor to use a Chart model class.
- Refactor ChartService so that it uses a ClimateModel and Scenario property rather than saving strings (which doesn't take advantage of typings). Probably makes sense to do this in combination with the previous TODO


## Testing Instructions

 * Restart dev server
 * Add a few indicators, set some options on them
 * Return to dashboard
 * Click back into the Lab view from the dashboard -- All previous state should be saved

Closes #54 and #69
